modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java

```diff
@@ -1101,24 +1101,55 @@ public static Schema unaliasSchema(OpenAPI openAPI,
                         return schema; // generate a model extending map
                     } else {
                         // treat it as a typical map
-                        return unaliasSchema(openAPI, allSchemas.get(ModelUtils.getSimpleRef(schema.get$ref())),
+                        return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())),
                                 importMappings);
                     }
                 }
             } else if (isObjectSchema(ref)) { // model
                 if (ref.getProperties() != null && !ref.getProperties().isEmpty()) { // has at least one property
                     return schema;
-                } else { // free form object (type: object)
-                    return unaliasSchema(openAPI, allSchemas.get(ModelUtils.getSimpleRef(schema.get$ref())),
+                } else {
+                    // free form object (type: object)
+                    if (hasValidation(ref)) {
+                        return schema;
+                    } else if (hasDescendants(simpleRef, openAPI)) {
+                        return schema;
+                    }
+                    return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())),
                             importMappings);
                 }
             } else {
-                return unaliasSchema(openAPI, allSchemas.get(ModelUtils.getSimpleRef(schema.get$ref())), importMappings);
+                return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())), importMappings);
             }
         }
         return schema;
     }
 
+    private static boolean hasDescendants(String simpleRef, OpenAPI openAPI) {
+        for (Map.Entry<String, Schema> entry : getSchemas(openAPI).entrySet()) {
+            if (!isSameSchema(simpleRef, entry.getKey())) {
+                Schema schema = entry.getValue();
+                if (isComposedSchema(schema)) {
+                    ComposedSchema composedChild = (ComposedSchema) schema;
+                    boolean b = composedChild.getAllOf()
+                            .stream()
+                            .map(Schema::get$ref)
+                            .filter(Objects::nonNull)
+                            .map(ModelUtils::getSimpleRef)
+                            .anyMatch(simpleRef::equals);
+                    if (b) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private static boolean isSameSchema(String simpleRef, String schema) {
+        return simpleRef.equals(schema);
+    }
```

Comment:
nit:
How about removing this function? It is one line and you can use `simpleRef.equals(schema)` in the call site instead. Also there is only one call site so it may not make sense to make this a function.

Reply:
I agree here. `isSameSchema(string, string)` doesn't give us any way to control drift of which strings are passed in here. This would better be suited as a comment above the conditional:

```
// check if model names are the same
if (!simpleRef.equals(entry.getKey())) {
```

Also, I wonder if checking `simpleRef` like this will work in all cases? For example, if I have `#/components/schemas/Cat` and `#/components/responses/Cat`, is there ever a situation where I'll see one `Cat` where I expect to see the other (see `ModelUtils.getSimpleRef`)?

Reply:
I've inlined the method. I don't manage to create such a case where I have a conflict where I have the wrong model because all entries came from `getSchemas`.